Rate limit ADR update (#951)
kyma-project/istio: docs/contributor/adr/0008-rate-limit.md

                        value: internal
                      - key: x-api-version-key
                        value: v2
                    token_bucket:
                      max_tokens: 60
                      tokens_per_fill: 60
                      fill_interval: 60s
```

### Enforcing Rate Limit

The [Envoy LocalRateLimit filter](https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/filters/http/local_ratelimit/v3/local_rate_limit.proto#local-rate-limit-proto)
allows to configure the enforcement for 0 - 100% percentage of the traffic by using the field **filter_enforced**.   
It is considered a good use case to allow to create a RateLimit CR without enforcing the rate limit. By following this
approach, it is possible to check the rate limit metrics to understand the impact on real traffic before blocking it.  
The decision is to hide the complexity of the Envoy enforcement configuration behind the optional boolean field
*spec.enforce* with the default value set to `true`.

### Rate Limit HTTP Response Headers

While enabling rate limit feature on the HTTP layer, there is a possibility to give insights to the client about the

> current rate limiting state e.g. how many requests are left to be done before being rate limited, or how much time is

Reviewer: ```suggestion
current rate limiting state, for example, how many requests are left to be done before being rate limited, or how much time is
```

left until rate limit tokens are refilled.  
The decision is that the user can enable rate limit headers in the RateLimit CR by setting the boolean field

> *spec.enableResponseHeaders*. Following security good practices, these headers are disabled by default to limit

Reviewer: ```suggestion
**spec.enableResponseHeaders**. Following security good practices, these headers are disabled by default to limit
```

| local.defaultBucket.fillInterval  | duration            | The fill interval that tokens are added to the bucket. <br/>During each fill interval, `tokensPerFill` are added to the bucket. The bucket will never contain more than `maxTokens` tokens. The `fillInterval` must be greater than or equal to 50ms to avoid excessive refills.                                                                                                                   | yes      |
| local.buckets                     | array               | List of additional rate limit buckets for requests. <br/>Each bucket must specify either a `path` or `headers`. <br/>For each request matching the bucket's criteria, a single token is consumed. If a token is available, the request is allowed. If no tokens are available, the request is rejected with status code `429`.                                                                     | no       |

> | local.buckets.path                | string              | Specifies the path to be rate limited starting with `/`. <br/>Example: "/foo"                                                                                                                                                                                                                                                                                                                      | no       |

Reviewer: ```suggestion
| local.buckets.path                | string              | Specifies the path to be rate limited starting with `/`. <br/>For example, `/foo`.                                                                                                                                                                                                                                                                                                                     | no       |
```

> | local.buckets.headers             | map<string, string> | Specifies the request headers to be rate limited. The key is the header name, and the value is the header value. All specified headers must be present in the request for this configuration to match. Example: "x-api-usage: BASIC".                                                                                                                                                              | no       |                                                                                                                                                                                    

Reviewer: ```suggestion
| local.buckets.headers             | map<string, string> | Specifies the request headers to be rate limited. The key is the header name, and the value is the header value. All specified headers must be present in the request for this configuration to match. For example, `x-api-usage: BASIC`.                                                                                                                                                              | no       |                                                                                                                                                                                    
```

| local.buckets.fillInterval        | duration            | The fill interval that tokens are added to the bucket. <br/>During each fill interval, `tokensPerFill` are added to the bucket. The bucket will never contain more than `maxTokens` tokens. The `fillInterval` must be greater than or equal to 50ms to avoid excessive refills.                                                                                                                   | yes      |
| enableResponseHeaders             | boolean             | Enables **x-rate-limit** response headers. The default value is `false`.                                                                                                                                                                                                                                                                                                                           | no       |

> | enforce                           | boolean             | Specifies whether the rate limit should be enforced. Default is `true`.                                                                                                                                                                                                                                                                                                                            | no       |

Reviewer: ```suggestion
| enforce                           | boolean             | Specifies whether the rate limit should be enforced. The default value is `true`.                                                                                                                                                                                                                                                                                                                            | no       |
```

![Istio Controller overview with RateLimit](../../assets/istio-operator-overview-ratelimit.svg)

## Consequences

New controller for the new RateLimit CRD needs to be implemented as a part of the Istio module. The new RateLimit CR can
be used to set rate limits in the cluster's service mesh without having to worry about possible changes in the Istio
EnvoyFilter resources. Nevertheless, EnvoyFilter is complex and since its API is not stable yet, the current RateLimit
CR will be provided in `v1alpha1` version and might be changed in the future Istio releases.    
The RateLimit CRD should be included in the blocking deletion strategy for Istio CR since Istio should not be
uninstalled while RateLimit CRs exist in the cluster.    
Requests beyond the allowed rate limit threshold will get the HTTP `429` response.

In general, it should not be necessary for users to create resources in the `istio-system` namespace. However, for rate
limiting the Istio Ingress Gateway, the RateLimit CR must be created in the `istio-system` namespace.

With RateLimit CR supporting multiple rate limit configuration entries it is hard to create accurate metrics for each of

> the rate limits. Per default the metrics are disabled and if enabled, the metrics will be aggregated, and it will be

Reviewer: ```suggestion
the rate limits. Per default, the metrics are disabled. If enabled, the metrics will be aggregated, and it will be
```
